community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NonUniqueNativeSchemaIndexPopulator.java

```diff
@@ -0,0 +1,109 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.neo4j.index.internal.gbptree.Layout;
+import org.neo4j.index.internal.gbptree.RecoveryCleanupWorkCollector;
+import org.neo4j.io.pagecache.PageCache;
+import org.neo4j.kernel.api.index.IndexEntryUpdate;
+import org.neo4j.kernel.impl.api.index.sampling.DefaultNonUniqueIndexSampler;
+import org.neo4j.kernel.impl.api.index.sampling.IndexSamplingConfig;
+import org.neo4j.kernel.impl.api.index.sampling.NonUniqueIndexSampler;
+import org.neo4j.storageengine.api.schema.IndexSample;
+
+/**
+ * {@link NativeSchemaIndexPopulator} which can accept duplicate values (for different entity ids).
+ */
+class NonUniqueNativeSchemaIndexPopulator<KEY extends SchemaNumberKey, VALUE extends SchemaNumberValue>
+        extends NativeSchemaIndexPopulator<KEY,VALUE>
+{
+    private final IndexSamplingConfig samplingConfig;
+    private boolean updateSampling;
+    private NonUniqueIndexSampler sampler;
+
+    NonUniqueNativeSchemaIndexPopulator( PageCache pageCache, File storeFile, Layout<KEY,VALUE> layout,
+            RecoveryCleanupWorkCollector recoveryCleanupWorkCollector, IndexSamplingConfig samplingConfig )
+    {
+        super( pageCache, storeFile, layout, recoveryCleanupWorkCollector );
+        this.samplingConfig = samplingConfig;
+    }
+
+    @Override
+    public void includeSample( IndexEntryUpdate update )
+    {
+        if ( updateSampling )
+        {
+            checkSampler();
+            sampler.include( SamplingUtil.encodedStringValuesForSampling( update.values() ) );
+        }
+    }
+
+    @Override
+    public void configureSampling( boolean onlineSampling )
+    {
+        this.updateSampling = onlineSampling;
+        this.sampler = onlineSampling ? new DefaultNonUniqueIndexSampler( samplingConfig.sampleSizeLimit() )
+                                      : new FullScanNonUniqueIndexSampler<>( tree, layout, samplingConfig );
+    }
+
+    @Override
+    public IndexSample sampleResult()
+    {
+        checkSampler();
+
+        // Close the writer before scanning
+        try
+        {
+            closeWriter();
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+
+        try
+        {
+            return sampler.result();
+        }
+        finally
+        {
+            // Start the writer again (TODO may be unnecessary)
```

Comment:
It's still a TODO although with better description now. This is what it now says:

```
            // Start the writer again (TODO may be unnecessary since this happens after all population is completed,
            // however would be weird to have this sampling method close the writer and tree and keep them closed.
            // Time will tell whether or not this could be avoided.
```